operator/src/main/java/oracle/kubernetes/operator/steps/InitializeInternalIdentityStep.java

```diff
@@ -0,0 +1,160 @@
+import java.io.StringWriter;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.security.Key;
+import java.security.KeyPair;
+import java.security.cert.X509Certificate;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+import io.kubernetes.client.custom.V1Patch;
+import io.kubernetes.client.openapi.models.V1ObjectMeta;
+import io.kubernetes.client.openapi.models.V1Secret;
+import jakarta.json.Json;
+import jakarta.json.JsonPatchBuilder;
+import oracle.kubernetes.operator.calls.CallResponse;
+import oracle.kubernetes.operator.helpers.CallBuilder;
+import oracle.kubernetes.operator.helpers.ResponseStep;
+import oracle.kubernetes.operator.logging.LoggingFacade;
+import oracle.kubernetes.operator.logging.LoggingFactory;
+import oracle.kubernetes.operator.utils.SelfSignedCertGenerator;
+import oracle.kubernetes.operator.work.NextAction;
+import oracle.kubernetes.operator.work.Packet;
+import oracle.kubernetes.operator.work.Step;
+import org.bouncycastle.openssl.jcajce.JcaPEMWriter;
+
+import static oracle.kubernetes.operator.helpers.NamespaceHelper.getOperatorNamespace;
+import static oracle.kubernetes.operator.logging.MessageKeys.INTERNAL_IDENTITY_INITIALIZATION_FAILED;
+import static oracle.kubernetes.operator.utils.Certificates.INTERNAL_CERTIFICATE;
+import static oracle.kubernetes.operator.utils.Certificates.INTERNAL_CERTIFICATE_KEY;
+import static oracle.kubernetes.operator.utils.SelfSignedCertGenerator.createKeyPair;
+
+public class InitializeInternalIdentityStep extends Step {
+
+  private static final LoggingFacade LOGGER = LoggingFactory.getLogger("Operator", "Operator");
+  private static final String OPERATOR_CM = "weblogic-operator-cm";
+  private static final String OPERATOR_SECRETS = "weblogic-operator-secrets";
+  private static final String SHA_256_WITH_RSA = "SHA256withRSA";
+  private static final String COMMON_NAME = "weblogic-operator";
+
+  public InitializeInternalIdentityStep(Step next) {
+    super(next);
+  }
+
+  @Override
+  public NextAction apply(Packet packet) {
+    try {
+      KeyPair keyPair = createKeyPair();
+      String key = convertToPEM(keyPair.getPrivate());
+      writeToFile(key, new File(INTERNAL_CERTIFICATE_KEY));
+      X509Certificate cert = SelfSignedCertGenerator.generate(keyPair, SHA_256_WITH_RSA, COMMON_NAME, 3650);
+      writeToFile(getBase64Encoded(cert), new File(INTERNAL_CERTIFICATE));
+      return doNext(recordInternalOperatorCert(cert,
+              recordInternalOperatorKey(key, getNext())), packet);
+    } catch (Exception e) {
+      LOGGER.severe(INTERNAL_IDENTITY_INITIALIZATION_FAILED, e.toString());
+      return doNext(getNext(), packet);
+    }
+  }
+
+  private static String convertToPEM(Key key) throws IOException {
+    StringWriter writer = new StringWriter();
+    JcaPEMWriter pemWriter = new JcaPEMWriter(writer);
+    pemWriter.writeObject(key);
+    pemWriter.flush();
+    return writer.toString();
+  }
```

Comment:
Fixed. I made the change to close the JcaPEMWriter. As per StringWriter java doc "Closing a StringWriter has no effect".

Reply:
Can you use try-with-resources pattern? https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html. This will close the stream even if something fails.